src/lib/provable/crypto/foreign-twisted.ts

```diff
@@ -0,0 +1,309 @@
+import { Provable } from '../provable.js';
+import { provableFromClass } from '../types/provable-derivers.js';
+import { l2Mask, multiRangeCheck } from '../gadgets/range-check.js';
+import { Bytes } from '../bytes.js';
+
+// external API
+export { createForeignTwisted, ForeignTwisted };
+
+// internal API
+export { toPoint, FlexiblePoint };
+
+type FlexiblePoint = {
+  x: AlmostForeignField | Field3 | bigint | number;
+  y: AlmostForeignField | Field3 | bigint | number;
+};
+function toPoint({ x, y }: ForeignTwisted): Point {
+  return { x: x.value, y: y.value };
+}
+
+class ForeignTwisted {
+  x: AlmostForeignField;
+  y: AlmostForeignField;
+
+  /**
+   * Create a new {@link ForeignTwisted} from an object representing the (twisted) x and y coordinates.
+   *
+   * Note: Inputs must be range checked if they originate from a different field with a different modulus or if they are not constants. Please refer to the {@link ForeignField} constructor comments for more details.
+   *
+   * @example
+   * ```ts
+   * let x = new ForeignTwisted({ x: 1n, y: 1n });
+   * ```
+   *
+   * **Warning**: This fails for a constant input which does not represent an actual point on the curve.
+   */
+  constructor(g: {
+    x: AlmostForeignField | Field3 | bigint | number;
+    y: AlmostForeignField | Field3 | bigint | number;
+  }) {
+    this.x = new this.Constructor.Field(g.x);
+    this.y = new this.Constructor.Field(g.y);
+    // don't allow constants that aren't on the curve
+    if (this.isConstant()) {
+      this.assertOnCurve();
+    }
+  }
+
+  /**
+   * Coerce the input to a {@link ForeignTwisted}.
+   */
+  static from(g: ForeignTwisted | FlexiblePoint) {
+    if (g instanceof this) return g;
+    return new this(g);
+  }
+
+  /**
+   * Parses a hexadecimal string representing an uncompressed elliptic curve point
+   * and coerces it into a {@link ForeignTwisted} point.
+   *
+   * The method extracts the x and y coordinates from the provided hex string and
+   * verifies that the resulting point lies on the curve.
+   *
+   * **Note:** This method only supports uncompressed elliptic curve points, which
+   * are 65 bytes in total (1-byte prefix + 32 bytes for x + 32 bytes for y).
+   *
+   * @param hex - The hexadecimal string representing the uncompressed elliptic curve point.
+   * @returns - A point on the foreign curve, parsed from the given hexadecimal string.
+   *
+   * @throws - Throws an error if the input is not a valid public key.
+   *
+   * @example
+   * ```ts
+   * class Edwards25519 extends createForeignCurveTwisted(Crypto.TwistedCurveParams.Edwards25519) {}
+   *
+   * // Example hex string for uncompressed point
+   * const publicKeyHex = '04f8b8db25c619d0c66b2dc9e97ecbafafae...';
+   * const point = Edwards25519.fromHex(publicKeyHex);
+   * ```
+   *
+   * **Important:** This method is only designed to handle uncompressed elliptic curve points in hex format.
+   */
+  static fromHex(hex: string) {
```

Comment:
why does this method exist? is this a common format used in applications? if yes that should be mentioned.

but in general I don't think packed hex strings are a great format that we should encourage people to use

Reply:
I thought it could be useful for testing purposes, just like in the weierstrass counterpart https://github.com/o1-labs/o1js/blob/c8240013102706c9113a0ccc4147a4a949a8cf29/src/lib/provable/crypto/foreign-curve.ts#L92